### ISSUE TITLE
Different training performance between mxnet v. 0.11.0 and v. 0.12.1

### ISSUE THREAD

**Issue by VGalata** [apache/incubator-mxnet#8792]
## Description

I get very different performance results during training for the same setup using `mxnet` versions `0.11.0` and `0.12.1`.

Example (link to the data is provided below):

- v. `0.11.0`:
    - CPU: improving performance, from 0.81 (epoch 1) to 0.98 (epoch 50)
    - GPU: improving performance, from 0.80 (epoch 1) to 0.98 (epoch 50)
- v. `0.12.1`: output changes though using `set.seed()`
    - CPU/GPU:
        - In some cases: fluctuating performance (got results between 0.4 and 0.7)
        - In other cases: only `NA` for each (printed) batch and epoch

## Environment info

```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              32
On-line CPU(s) list: 0-31
Thread(s) per core:  2
Core(s) per socket:  16
Socket(s):           1
NUMA node(s):        1
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               1
Model name:          AMD Ryzen Threadripper 1950X 16-Core Processor
Stepping:            1
CPU MHz:             3393.339
CPU max MHz:         3400,0000
CPU min MHz:         2200,0000
BogoMIPS:            6786.67
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           64K
L2 cache:            512K
L3 cache:            8192K
NUMA node0 CPU(s):   0-31
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid amd_dcm aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_l2 mwaitx hw_pstate vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 xsaves clzero irperf arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload overflow_recov succor smca
----------Python Info----------
('Version      :', '2.7.14')
('Compiler     :', 'GCC 7.2.0')
('Build        :', ('default', 'Sep 23 2017 22:06:14'))
('Arch         :', ('64bit', ''))
------------Pip Info-----------
('Version      :', '9.0.1')
('Directory    :', '/usr/lib/python2.7/dist-packages/pip')
----------MXNet Info-----------
No MXNet installed.
----------System Info----------
('Platform     :', 'Linux-4.13.0-17-generic-x86_64-with-Ubuntu-17.10-artful')
('system       :', 'Linux')
('node         :', 'blackbox')
('release      :', '4.13.0-17-generic')
('version      :', '#20-Ubuntu SMP Mon Nov 6 10:04:08 UTC 2017')
----------Hardware Info----------
('machine      :', 'x86_64')
('processor    :', 'x86_64')
----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.0078 sec, LOAD: 0.8450 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0068 sec, LOAD: 0.3261 sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0245 sec, LOAD: 0.3651 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0113 sec, LOAD: 0.0449 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.0331 sec, LOAD: 0.0431 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.0240 sec, LOAD: 0.7559 sec.
```

Package used: R-package, versions `0.11.0` and `0.12.1` compiled for GPU usage.

R `sessionInfo()` for `0.11.0`

```
R version 3.4.2 (2017-09-28)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 17.10

Matrix products: default
BLAS: /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] methods   stats     graphics  grDevices utils     datasets  base

other attached packages:
[1] mxnet_0.11.0        matrixStats_0.52.2  data.table_1.10.4-3
[4] argparse_1.0.7      proto_1.0.0         testit_0.7
[7] futile.logger_1.4.3

loaded via a namespace (and not attached):
 [1] Rook_1.1-1           purrr_0.2.4          colorspace_1.3-2
 [4] htmltools_0.3.6      getopt_1.20.0        viridisLite_0.2.0
 [7] XML_3.98-1.9         rlang_0.1.4          glue_1.2.0
[10] RColorBrewer_1.1-2   lambda.r_1.2         bindrcpp_0.2
[13] bindr_0.1            plyr_1.8.4           stringr_1.2.0
[16] munsell_0.4.3        findpython_1.0.2     gtable_0.2.0
[19] visNetwork_2.0.1     htmlwidgets_0.9      codetools_0.2-15
[22] DiagrammeR_0.9.2     Rcpp_0.12.13         readr_1.1.1
[25] scales_0.5.0         jsonlite_1.5         rgexf_0.15.3
[28] gridExtra_2.3        brew_1.0-6           rjson_0.2.15
[31] ggplot2_2.2.1        hms_0.3              digest_0.6.12
[34] stringi_1.1.6        dplyr_0.7.4          grid_3.4.2
[37] influenceR_0.1.0     tools_3.4.2          magrittr_1.5
[40] lazyeval_0.2.1       tibble_1.3.4         futile.options_1.0.0
[43] tidyr_0.7.2          pkgconfig_2.0.1      downloader_0.4
[46] assertthat_0.2.0     rstudioapi_0.7       viridis_0.4.0
[49] R6_2.2.2             igraph_1.1.2         compiler_3.4.2
```

R `sessionInfo()` for `0.12.1`

```
R version 3.4.2 (2017-09-28)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 17.10

Matrix products: default
BLAS: /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] methods   stats     graphics  grDevices utils     datasets  base

other attached packages:
[1] mxnet_0.12.1        matrixStats_0.52.2  data.table_1.10.4-3
[4] argparse_1.0.7      proto_1.0.0         testit_0.7
[7] futile.logger_1.4.3

loaded via a namespace (and not attached):
 [1] Rook_1.1-1           purrr_0.2.4          colorspace_1.3-2
 [4] htmltools_0.3.6      getopt_1.20.0        viridisLite_0.2.0
 [7] XML_3.98-1.9         rlang_0.1.4          glue_1.2.0
[10] RColorBrewer_1.1-2   lambda.r_1.2         bindrcpp_0.2
[13] bindr_0.1            plyr_1.8.4           stringr_1.2.0
[16] munsell_0.4.3        findpython_1.0.2     gtable_0.2.0
[19] visNetwork_2.0.1     htmlwidgets_0.9      codetools_0.2-15
[22] DiagrammeR_0.9.2     Rcpp_0.12.13         readr_1.1.1
[25] scales_0.5.0         jsonlite_1.5         rgexf_0.15.3
[28] gridExtra_2.3        brew_1.0-6           rjson_0.2.15
[31] ggplot2_2.2.1        hms_0.3              digest_0.6.12
[34] stringi_1.1.6        dplyr_0.7.4          grid_3.4.2
[37] influenceR_0.1.0     tools_3.4.2          magrittr_1.5
[40] lazyeval_0.2.1       tibble_1.3.4         futile.options_1.0.0
[43] tidyr_0.7.2          pkgconfig_2.0.1      downloader_0.4
[46] assertthat_0.2.0     rstudioapi_0.7       viridis_0.4.0
[49] R6_2.2.2             igraph_1.1.2         compiler_3.4.2
```

## Build info

Compiler: `gcc (Ubuntu 7.2.0-8ubuntu3) 7.2.0`

MXNet commit hash:

- `a5edbf94094581ee27157eae4f2113115a3994e7` for `0.11.0`
- `e0c7906693f0c79b0ce34a4d777c26a6bf1903c1` for `0.12.1`

Build config: (Linux, R-package, GPU)

```bash
git clone --recursive https://github.com/apache/incubator-mxnet.git mxnet --branch <version>
cd mxnet
make -j $(nproc) USE_OPENCV=1 USE_BLAS=openblas USE_CUDA=1 USE_CUDA_PATH=/usr/local/cuda USE_CUDNN=1
make rpkg
R CMD INSTALL mxnet_current_r.tar.gz
```

where `<version>` was `0.11.0` and `0.12.1` respectively.

## Minimum reproducible example

[Link to data containing an example (ZIP archive)](https://oc-ccb.cs.uni-saarland.de/owncloud/index.php/s/pai4qKubRqoHXHB)

- `train_array.rds`: Training data
    - 8160 samples, 512 features (zero mean, unit variance)
    - Set dimensions: 1 x 512 x 1 x 8160
- `train_labels.rds`: Class labels - binary (0/1), 2622 are 0, 5538 are 1
- `network.R`: Network structure
- `run_mxnet.R`: Script to load data and start training

## Steps to reproduce

```bash
Rscript run_mxnet.R "<R_lib_path>" "<cpu|gpu>"
```

where `R_lib_path` is the location of the `mxnet` R-package to be used.

**Comment by jeremiedb** [comment]
There's been a bug introduced for a short time in master after `mx.runif` and `mx.rnorm` used by the initializer stopped working. It has been fixed 2 weeks ago by commit #8121 by using the random generator from the ndarray api (`mx.nd.random.uniform`). Building from the current master version should work fine.

**Comment by VGalata** [comment]
Dear @jeremiedb,

Thank you for your fast reply!

I downloaded the `master` branch (`v0.9.3a-1486-g12643131`, commit hash `1264313183d35c86ed49e7ec708a076fe858325f`) and built the package. The results look much better now, they are comparable to the output of version `0.11.0`. But, using `set.seed()` has no effect on the result reproducibility, i.e. running the same script multiple times generates different results.

**Comment by jeremiedb** [comment]
I think the reproducibility issue is solely due to the dropout layer. If you run a model without dropout, you should get reproducible results. I'm not sure though how to avoid this behavior other than by creating custom masks.

**Comment by VGalata** [comment]
Hm... It does not seem to be due to the dropout layers. Using version `0.11.0` gives the same accuracy values with or without the dropout layers. For the `master` branch version mentioned above, the results also vary when I remove the dropout layers from my example network.

**Comment by eric-haibin-lin** [comment]
related issue: https://github.com/apache/incubator-mxnet/issues/7426

**Comment by bhavinthaker** [comment]
Proposed labels: "performance", "R"

**Comment by anirudhacharya** [comment]
@VGalata I tried with the latest v1.3( source build) and was not able to reproduce this issue with the example you provided. Can you please verify and reopen the issue if problem persists.

@nswamy please close this issue.

**Comment by nswamy** [comment]
@VGalata Closing this issue.
@anirudhacharya I think @anirudhacharya meant he tried from the master(v1.3 is coming out soon). Please create a new Issue if you find issues.
